x/gamm/keeper/grpc_query.go

```diff
@@ -140,6 +140,72 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
 	}, err
 }
 
+// CalcJoinPoolShares queries the amount of shares you get by providing specific amount of tokens
+func (q Querier) CalcJoinPoolShares(ctx context.Context, req *types.QueryCalcJoinPoolSharesRequest) (*types.QueryCalcJoinPoolSharesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	if req.TokensIn == nil {
+		return nil, status.Error(codes.InvalidArgument, "no tokens in")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.getPoolForSwap(sdkCtx, req.PoolId)
+	if err != nil {
+		return nil, err
+	}
+
+	numShares, newLiquidity, err := pool.CalcJoinPoolShares(sdkCtx, req.TokensIn, pool.GetSwapFee(sdkCtx))
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.QueryCalcJoinPoolSharesResponse{
+		ShareOutAmount: numShares,
+		TokensOut:      newLiquidity,
+	}, nil
+}
+
+// CalcExitPoolCoinsFromShares queries the amount of tokens you get by exiting specific amouint of shares
+func (q Querier) CalcExitPoolCoinsFromShares(ctx context.Context, req *types.QueryCalcExitPoolCoinsFromSharesRequest) (*types.QueryCalcExitPoolCoinsFromSharesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.GetPoolAndPoke(sdkCtx, req.PoolId)
+	if err != nil {
+		return nil, types.ErrPoolNotFound
+	}
+
+	exitFee := pool.GetExitFee(sdkCtx)
+
+	totalSharesAmount := pool.GetTotalShares()
+	if req.ShareInAmount.GTE(totalSharesAmount) || req.ShareInAmount.LTE(sdk.ZeroInt()) {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "share ratio is zero or negative")
+	}
+
+	exitCoins, err := pool.CalcExitPoolCoinsFromShares(sdkCtx, req.ShareInAmount, exitFee)
+	if err != nil {
+		return nil, err
+	}
+
+	coinsOut := sdk.NewCoins()
+	for _, req_denom := range req.TokensOutDenoms {
```

Comment:
I'm slightly confused about the purpose of `TokensOutDenoms` for filtering. Is this a requirement for this query?

Why can't we let clients filter out the denoms that they want? I think that filtering out by this input here is likely to lead to erroneous use where clients provide only one `TokensOutDenoms` while there are more existing

Reply:
Agreed, let me remove that parameter and remove this for loop.

Reply:
> I'm slightly confused about the purpose of `TokensOutDenoms` for filtering. Is this a requirement for this query?
> 
> Why can't we let clients filter out the denoms that they want? I think that filtering out by this input here is likely to lead to erroneous use where clients provide only one `TokensOutDenoms` while there are more existing

I've removed this parameter and fixed the typos that you also suggested.